### ISSUE TITLE
Block the pull-request merge depending on sonar issues

### ISSUE THREAD

**Issue by juherr** [AmadeusITGroup/sonar-stash#11]
I'd like to block the merge of pull-request if some sonar issues are created.
The rule should be configurable. For example: "no blocker/critical/major, not more than 5 minor/info".

I'm not sure `sonar-stash` will be able to do that and maybe it should be done by a stash plugin, notified by `sonar-stash`.

**Comment by acopet** [comment]
I am not sure this feature needs to be developed in the sonar-stash plugin.
For me, the build breaker plugin is on charge to fail a build if a project does not pass a SQ quality gate. And Stash can avoid the merge of a pull-request if the build status is failure...

**Comment by juherr** [comment]
I wasn't aware of it but the doc says:

> Starting with SonarQube 5.1, the Build Breaker plugin does not work any longer in the preview & incremental modes

**Comment by juherr** [comment]
I confirm: I didn't succeed to use the build breaker plugin in incremental mode.

But even if it works, something (sonar-stash?) should describe in the PR UI why the job failed (if the reason is due to the quality gate).

What do you think?

**Comment by juherr** [comment]
I found some answers: https://groups.google.com/forum/#!topic/sonarqube/gR07SqqpAyQ

**Comment by acopet** [comment]
Finally, just have SonarQube as a real approver of the pull-request could fit our needs: if we reach the expected threshold, SonarQube automatically approves the pull-request. And merge condition will stay in charge of Stash.

**Comment by juherr** [comment]
:+1: I like this idea

**Comment by christiangalsterer** [comment]
Unfortunately the Stash/Bitbucket REST API doesn't provide an endpoint to veto against a PR.
This functionality is currently only available as a MergeCheckHook where one can then use the Stash API to veto against a PR.

**Comment by acopet** [comment]
**Pull-request #27 submitted.**

For this first version, SonarQube user only approves the pull-request if there is no new issue raised by the pull-request. All the mechanism is in place with this pull-request: we will develop more complex approval rules later through another ticket.

This new feature is optional and can be deactivated through plugin configuration page (and so SQ options).

Cheers,
Antoine

**Comment by juherr** [comment]
:+1: When do you plan to release it?

**Comment by acopet** [comment]
We have planned to release this ticket through the next plugin version (1.1.0).
This release will contain others tickets like issue #14.

**Comment by acopet** [comment]
**Feature available on master branch and will be release with plugin version 1.1.0.**

**Comment by hax0rdlux3** [comment]
Hello,

 Has any progress been made on preventing/proceeding with merge depending on issues found by SQ for a specific pull request? We are hoping to have pull requests approved and merged automatically if no issues below a specified severity are set. 

thanks!

**Comment by t-8ch** [comment]
@hax0rdlux3 Currently only the approval ist implemented.
I am not sure if we want to have the merging happen by the sonar-stash plugin or it should be handled by Bitbucket itself.

**Comment by hax0rdlux3** [comment]
Understood... I'll pursue other implementation options. I appreciate the reply.
